File: components/gitpod-protocol/src/usage.ts
/**
 * Copyright (c) 2022 Gitpod GmbH. All rights reserved.
 * Licensed under the GNU Affero General Public License (AGPL).
 * See License.AGPL.txt in the project root for license information.
 */

import { WorkspaceType } from "./protocol";

// types below are manually kept in sycn with components/usage-api/typescript/src/usage/v1/usage_pb.d.ts
export interface ListUsageRequest {
    attributionId: string;
    userId?: string;
    from?: number;
    to?: number;
    order: Ordering;
    pagination?: PaginationRequest;
}

export enum Ordering {
    ORDERING_DESCENDING = 0,
    ORDERING_ASCENDING = 1,
}

export interface PaginationRequest {
    perPage: number;
    page: number;
}

export interface ListUsageResponse {
    usageEntriesList: Usage[];
    pagination?: PaginationResponse;
    creditsUsed: number;
    ledgerIntervalMinutes?: number;
}

export interface PaginationResponse {
    perPage: number;
    totalPages: number;
    total: number;
    page: number;
}

export type UsageKind = "workspaceinstance" | "invoice";
export interface Usage {
    id: string;
    attributionId: string;
    description: string;
    credits: number;
    effectiveTime?: number;
    kind: UsageKind;
    workspaceInstanceId: string;
    draft: boolean;
    metadata: WorkspaceInstanceUsageData | InvoiceUsageData;
}

// the equivalent golang shape is maintained in `/workspace/gitpod/`components/usage/pkg/db/usage.go`
export interface WorkspaceInstanceUsageData {
    workspaceId: string;
    workspaceType: WorkspaceType;
    workspaceClass: string;
    contextURL: string;
    creationTime?: string;
    startTime: string;
    endTime?: string;
    stoppedTime?: string;
    userId: string;
    userName: string;
    userAvatarURL: string;
}

export interface InvoiceUsageData {
    invoiceId: string;
    startDate: string;
    endDate: string;
}

export interface CostCenterJSON {
    attributionId: string;
    spendingLimit: number;
    billingStrategy: CostCenter_BillingStrategy;
    nextBillingTime?: string;
    billingCycleStart?: string;
}

export enum CostCenter_BillingStrategy {
    BILLING_STRATEGY_STRIPE = "BILLING_STRATEGY_STRIPE",
    BILLING_STRATEGY_OTHER = "BILLING_STRATEGY_OTHER",
    UNRECOGNIZED = "UNRECOGNIZED",
}

File: components/usage/pkg/db/dbtest/usage.go
// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License-AGPL.txt in the project root for license information.

package dbtest

import (
	"testing"

	"github.com/gitpod-io/gitpod/usage/pkg/db"
	"github.com/google/uuid"
	"github.com/stretchr/testify/require"
	"gorm.io/gorm"
)

func NewUsage(t *testing.T, record db.Usage) db.Usage {
	t.Helper()

	result := db.Usage{
		ID:                  uuid.New(),
		AttributionID:       db.NewUserAttributionID(uuid.New().String()),
		Description:         "some description",
		CreditCents:         42,
		EffectiveTime:       db.VarcharTime{},
		Kind:                "workspaceinstance",
		WorkspaceInstanceID: uuid.New(),
	}

	if record.ID.ID() != 0 {
		result.ID = record.ID
	}
	if record.EffectiveTime.IsSet() {
		result.EffectiveTime = record.EffectiveTime
	}
	if record.AttributionID != "" {
		result.AttributionID = record.AttributionID
	}
	if record.Description != "" {
		result.Description = record.Description
	}
	if record.CreditCents != 0 {
		result.CreditCents = record.CreditCents
	}
	if record.WorkspaceInstanceID.ID() != 0 {
		result.WorkspaceInstanceID = record.WorkspaceInstanceID
	}
	if record.Kind != "" {
		result.Kind = record.Kind
	}
	if record.Draft {
		result.Draft = true
	}
	if record.Metadata != nil {
		result.Metadata = record.Metadata
	}
	return result
}

func CreateUsageRecords(t *testing.T, conn *gorm.DB, entries ...db.Usage) []db.Usage {
	t.Helper()

	var records []db.Usage
	var ids []string
	for _, usageEntry := range entries {
		record := NewUsage(t, usageEntry)
		records = append(records, record)
		ids = append(ids, record.ID.String())
	}

	require.NoError(t, conn.CreateInBatches(&records, 1000).Error)

	t.Cleanup(func() {
		require.NoError(t, conn.Where(ids).Delete(&db.Usage{}).Error)
	})

	t.Logf("stored %d", len(entries))

	return records
}
